net: pkt: fix race condition in packet reference counting

dgloeck opened the pull request:
It has been observed that some network drivers, f.ex. the SAM E70 GMAC,
call net_pkt_unref from inside the interrupt that signals the successful
transmission of a packet. This conflicts with the net_pkt_unref call made
by ethernet_send after the packet has been given to the driver.

We fix this by using an atomic_t to hold the reference count as there
might be other, difficult to find cases of net_pkt_(un)ref being used
across threads and interrupts.

The name of the element has been changed from "ref" to "atomic_ref" to
cause a compile error when code still has not been converted to use the
atomic_* functions.

Fixes #12708

Signed-off-by: Daniel Glöckner <dg@emlix.com>

dgloeck committed:
net: pkt: fix race condition in packet reference counting

It has been observed that some network drivers, f.ex. the SAM E70 GMAC,
call net_pkt_unref from inside the interrupt that signals the successful
transmission of a packet. This conflicts with the net_pkt_unref call
made by ethernet_send after the packet has been given to the driver.

We fix this by using an atomic_t to hold the reference count as there
might be other, difficult to find cases of net_pkt_(un)ref being used
across threads and interrupts.

The name of the element has been changed from "ref" to "atomic_ref" to
cause a compile error when code still has not been converted to use the
atomic_* functions.

Fixes #12708

Signed-off-by: Daniel Glöckner <dg@emlix.com>

dgloeck commented:
Maybe I should note that I did only test this with the gPTP sample.
There might still be code that has to be converted from `ref` to `atomic_ref` that is not built with this sample.

tbursztyka commented:
seems like the wrong approach. The issue is known, see pr #12563, @aurel32 is working on the sam 70 driver side.

dgloeck commented:
@tbursztyka, I don't see how this is related to #12563.
It is true that my problem will no longer exist if #12563 ends up getting rid of the net_pkt_unref in tx_complete, but the core problem is that the three ARM assembly instructions that are generated from the `--pkt->ref` line in net_pkt_unref are interrupted by another high priority chunk of code that also wants to change the value of pkt->ref.

tbursztyka commented:
Your problem is that the driver does not ref count once more the necessary part. It's totally fine for net_pkt_unref() to be interrupted (the pkt flow from l3 -> l2 -> driver allows that). There is no need to make the ref atomic so far.

tbursztyka commented on subsys/net/ip/net_pkt.c:
```diff
@@ -809,20 +815,25 @@ struct net_pkt *net_pkt_ref_debug(struct net_pkt *pkt, const char *caller,
 struct net_pkt *net_pkt_ref(struct net_pkt *pkt)
 #endif /* NET_LOG_LEVEL >= LOG_LEVEL_DBG */
 {
-	if (!pkt) {
```
you want to keep that test, as for unref btw or it will enable dereferencing a NULL pointer at line 821

dgloeck replied:
I did keep that test by doing
```
ref = pkt ? atomic_get(&pkt->atomic_ref) : 0;
if (!ref) {
```
This additionally catches the error of passing a pkt with reference count of 0. Since the NET_ERR message prints pkt, the two errors can be distinguished.

pfalcon replied:
> This additionally catches the error of passing a pkt with reference count of 0.

Though if this done to catch 2 errors with one error message, maybe it's ok... But I'd suggest to add code comments for such cases.

dgloeck replied:
Efficiency should not be an issue. This loop is almost never executed more than once. Only when it is interrupted by another thread/interrupt handler, it has to loop.

I can spice up this function with `unlikely()` if you want. Both `!ref` and `!atomic_cas(&pkt->atomic_ref, ref, ref + 1)` are very unlikely.

tbursztyka reviewed (changes requested):
Nevermind my earlier comments, this is a necessary fix. Just make sure it passes CI.

dgloeck committed:
net: pkt: convert remaining users of ref to atomic_ref

The previous commit replaced the net_pkt element ref with an element
atomic_ref. CI tests turned up more places where ref was used directly.
This commit converts them to use the new element.

Signed-off-by: Daniel Glöckner <dg@emlix.com>

jukkar reviewed (approved):
LGTM

pfalcon commented:
@dgloeck, I wish you avoided making not directly related whitespace changes in this patch. This is now a massive patch which is very hard to review for correctness.

dgloeck commented:
@pfalcon, as far as I can see, there are no unrelated whitespace changes. A few lines of net_pkt_ref and net_pkt_unref had to be reindented because of an additional outer loop. Apart from that I had to split some lines that went above the 80 characters per line limit imposed by checkpatch.

pfalcon commented on include/net/net_pkt.h:
```diff
@@ -83,6 +83,9 @@ struct net_pkt {
 	u8_t *next_hdr;	/* where is the next header */
 
+	/** Reference counter */
+	atomic_t atomic_ref;
```
Ok, dumb question, "The name of the element has been changed from "ref" to "atomic_ref" to
cause a compile error" - sounds good, but why it was moved around?

dgloeck replied:
Because there it is guaranteed to be aligned to 4 bytes (assuming pointers are 4 bytes).
There where it was, it would have resulted in padding depending on CONFIG_NET_STATISTICS and CONFIG_NET_TCP.

pfalcon replied:
> Because there it is guaranteed to be aligned to 4 bytes

Right, as it changes from u8_t to atomic word. Ok.

pfalcon reviewed (approved):
Ok, I traced thru this, and believe I understand what it does now, and I don't see issues with that (but I'm not an atomic primitives expert, I'd just use irq_lock :-D). Approving.

pfalcon commented:
Oh, I guess I'd after all suggest to squash commits together - the first one has already enough green/red, so you can't spoil it much, and we'd win bisect buildability.

dgloeck commented:
@pfalcon, do you still want me to add that comment in net_pkt_ref and to squash the commits together? It would invalidate the four approvals.

pfalcon commented:
> do you still want me to add that comment in net_pkt_ref and to squash the commits together? It would invalidate the four approvals.

Comment should be safe and helpful. As for squashing, fairly speaking, I don't remember seeing automatic invalidation of approvals on Github for a while (I remember times when those were common). I would even think that maybe Github disabled such invalidation. But up to you to decide on that point.

pfalcon commented:
@dgloeck, Thinking about your patch more, I wonder if we now should have net_pkt_get_refcount()? @jukkar, @tbursztyka: What do you think?

dgloeck commented:
@pfalcon Do you believe we will have to touch the reference counting code again in the future? Then wrapping the access in a function would make sense.

But I don't think we should help people read the reference count. It might be outdated before they do something with the value. The only safe way to check if the refcount for a packet is zero is to look at the return value of net_pkt_ref. And if it was zero, net_pkt_ref will already have logged an error (if enabled).

pfalcon commented:
@dgloeck 

> Do you believe we will have to touch the reference counting code again in the future?

Yes. The whole idea of reference count is to allow for optimizations. Basic and obvious optimization is to avoid cloning a packet, instead having refcount >1. But having such a refcount value precludes other interesting optimizations. But our current situation is that in many (?, some, for sure) places there's implicit assumption that refcount == 1. See for example my own cute code: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/sockets/sockets.c#L606 (yes, that destructively modifies a packet). Such cases need to be fixed of course.

> It might be outdated before they do something with the value. The only safe way to check if the refcount for a packet is zero is to look at the return value of net_pkt_ref.

Per above, I'm interested in checking for refcount == 1. And surely, that would require some discipline. E.g., if a packet is destined for 2 separate processing pipelines, then refcount somehow needs to set to 2 before any of those pipelines "starts to run". That's why I'm consulting with you to detect possible issues beforehand.

> Then wrapping the access in a function would make sense.

I'll prepare a patch.

@rveerama1: FYI, That's related to your comment https://github.com/zephyrproject-rtos/zephyr/pull/12564#discussion_r253480160 . I'm preparing a response to it too.
